conduit/data/datasets/base.py

```diff
@@ -21,23 +32,42 @@
 )
 from conduit.logging import init_logger
 
-__all__ = ["CdtDataset"]
+__all__ = [
+    "CdtDataset",
+    "I",
+    "S",
+    "X",
+    "Y",
+]
+
+
+I = TypeVar(
+    "I",
+    NamedSample[LoadedData],
+    BinarySample[LoadedData],
+    SubgroupSample[LoadedData],
+    TernarySample[LoadedData],
+)
+
+X = TypeVar("X", bound=UnloadedData)
+S = TypeVar("S", bound=Optional[Tensor])
+Y = TypeVar("Y", bound=Optional[Tensor])
 
 
-class CdtDataset(DatasetProt[NamedSample]):
+class CdtDataset(DatasetProt[I], Generic[I, X, Y, S]):
     _repr_indent: ClassVar[int] = 4
     _logger: Optional[logging.Logger] = None
 
     def __init__(
-        self, *, x: UnloadedData, y: Optional[TargetData] = None, s: Optional[TargetData] = None
+        self, *, x: X, y: Optional[TargetData] = None, s: Optional[TargetData] = None
```

Comment:
I assume there is a reason why this is not
```suggestion
        self, *, x: X, y: Y = None, s: S = None
```
?

Reply:
Is this because `Y = Optional[Tensor]`, not `Y = Optional[TargetData]`?

Reply:
This has to do with the fact that numpy arrays are converted into tensors in the init